### ISSUE TITLE
YJIT: make sure we don't make the performance of unit tests worse

### ISSUE THREAD

**Issue by maximecb** [Shopify/ruby#449]
This is not a high priority issue but a nice-to-have. A current flaw with YJIT is that, while we manage speedups on almost every workload, we get a slight performance drop while running unit tests. This is because unit tests run a lot of code not very many times. That's unfortunate because it might negatively affect people's perception of YJIT. Unit tests are typically fairly long to run, so people might naturally be tempted to run YJIT there, and then be disappointed.

The first question is: how slow are we on unit tests? Should we set up a unit test benchmark? Should we simply use first iteration time on the benchmarks we already have as a proxy?

Fixing the problem could be as simple as increasing the default YJIT call threshold. It's currently set to 10, which is very aggressive, but it could easily be bumped to 50 or 100. We would still ideally want some kind of metric on which to base the choice of call threshold that we pick.

The downside of increasing the call threshold is that this could affect our benchmarks. Noah suggests that we could just manually set the benchmarks to use a call threshold of 10 to ensure that they warm up fast as they do now. That's probably acceptable. If we do that, we should probably do the same for yjit-bench.

@noahgibbs

**Comment by k0kubun** [comment]
> The downside of increasing the call threshold is that this could affect our benchmarks. Noah suggests that we could just manually set the benchmarks to use a call threshold of 10 to ensure that they warm up fast as they do now. 

I think we should keep using what most users would use: the actual default call threshold. When you change the threshold, the code layout/sequence could be impacted, and we should measure what actually happens to users and Shopify rather than artificially keeping numbers. 

If our warmup is unrealistic, we should just fix it. If it's already realistic, we should benchmark and make sure that the default gives the best performance to most users.

**Comment by noahgibbs** [comment]
If we raise the default threshold, we're definitely suggesting that YJIT is mostly for optimising long-running workloads (e.g. servers.) That's fine, but it also affects our other communication. For instance, very short workloads are just not going to be optimised. That's great for gem/bundler -- it means turning on YJIT for everything won't drop their performance significantly. But it means YJIT will want tuning for workloads it currently optimised.

I don't think there's a single right answer. But I think whatever tradeoff we pick, we should communicate clearly to users.

I don't care if we raise the threshold or not. Turning on YJIT by default is going to be error-prone no matter how we do it. But communication will be important. Luckily, the people that care the most are thrilled to have settings to play with. Some day when we think we have good default settings we'll want to write a nice tuning-guide blog post and summarise it for the docs :-)

**Comment by maximecb** [comment]
> If we raise the default threshold, we're definitely suggesting that YJIT is mostly for optimising long-running workloads (e.g. servers.) That's fine, but it also affects our other communication. For instance, very short workloads are just not going to be optimised. That's great for gem/bundler -- it means turning on YJIT for everything won't drop their performance significantly. But it means YJIT will want tuning for workloads it currently optimised.

The tradeoff with every JIT compiler is that the optimizations have to pay for themselves.

IMO the current value of 10 is kind of unrealistically low. I think what we'll find is that if we increase it to 20 for example, performance we get on benchmarks would actually not change much. This is because a lot of things in the way programs execute follow a Pareto distribution. A small minority of methods does most of the work, and these methods get executed a lot more than the rest.

I think we're doing a pretty good job on relatively short workloads already (a few seconds) and bumping the call threshold won't change that, but it does affect the way we benchmark a bit.

> I think we should keep using what most users would use: the actual default call threshold. When you change the threshold, the code layout/sequence could be impacted, and we should measure what actually happens to users and Shopify rather than artificially keeping numbers.

Kokubun makes a fair point that we should probably benchmark with the default parameters. In all honesty, it might be interesting to bump the call threshold to 20 and see what happens in YJIT bench. I might try doing that this afternoon 🤔 

> If our warmup is unrealistic, we should just fix it. If it's already realistic, we should benchmark and make sure that the default gives the best performance to most users.

Right now the way we do warmup is to run the benchmarks for a fixed number of warm up iterations. I insisted on doing that way because I think it's more predictable than a dynamic warm up system that tries to look at how long it takes to reach peak performance. TruffleRuby does it that way, and it takes them a very very long time to warm up, which IMO is also not at all how end-users use software. We also have the constraint that we want benchmarks to run within a reasonable amount of time. It's a complex problem, but I want to be open to other ideas as well.

**Comment by maximecb** [comment]
Out of curiosity, I decided to spin up an AWS x86 instance and benchmark with call threshold 10 vs 100 and also 10 vs 500 using `yjit-bench`. I left the warmup the same as before.

Interpreter vs call threshold 10 vs 100:
```
-------------  -----------  ----------  --------------  ----------  ---------------  ----------  ----------------  -----------------  -----------------  ------------------
bench          interp (ms)  stddev (%)  yjit_ct10 (ms)  stddev (%)  yjit_ct100 (ms)  stddev (%)  interp/yjit_ct10  yjit_ct10 1st itr  interp/yjit_ct100  yjit_ct100 1st itr
activerecord   149.9        0.4         99.4            0.6         98.9             0.7         1.51              1.20               1.52               1.17
erubi_rails    26.6         2.8         19.0            4.1         18.6             4.2         1.40              0.58               1.43               0.52
hexapdf        3106.5       1.2         2103.7          1.1         2167.3           1.7         1.48              1.34               1.43               1.31
liquid-render  194.8        0.7         116.5           1.4         115.3            1.3         1.67              1.30               1.69               1.31
mail           175.0        1.3         151.8           0.4         151.6            0.1         1.15              0.81               1.15               0.81
optcarrot      6327.0       0.7         3446.5          0.7         3484.3           0.6         1.84              1.79               1.82               1.77
psych-load     2419.1       0.0         1898.9          0.0         1908.1           0.0         1.27              1.27               1.27               1.26
railsbench     2766.0       0.8         2069.7          1.3         2051.7           1.1         1.34              1.21               1.35               1.22
rubykon        12556.5      0.4         6950.5          0.4         6922.3           0.6         1.81              1.83               1.81               1.84
-------------  -----------  ----------  --------------  ----------  ---------------  ----------  ----------------  -----------------  -----------------  ------------------
```

From what I see above there's surprisingly little performance difference between call threshold 10 and 100, even on the first iteration, which is surprising to me.

Interpreter vs call threshold 10 vs 500:
```
./run_benchmarks.rb -e "interp::ruby" -e "yjit_ct10::ruby --yjit --yjit-call-threshold=10" -e "yjit_ct500::ruby --yjit --yjit-call-threshold=500" \
activerecord \
hexapdf \
liquid-render \
mail \
psych-load \
railsbench \
erubi_rails \
rubykon \
optcarrot
-------------  -----------  ----------  --------------  ----------  ---------------  ----------  ----------------  -----------------  -----------------  ------------------
bench          interp (ms)  stddev (%)  yjit_ct10 (ms)  stddev (%)  yjit_ct500 (ms)  stddev (%)  interp/yjit_ct10  yjit_ct10 1st itr  interp/yjit_ct500  yjit_ct500 1st itr
activerecord   149.9        0.4         99.7            0.7         97.3             0.9         1.50              1.19               1.54               1.10
erubi_rails    26.5         2.6         19.0            3.9         18.3             3.7         1.40              0.55               1.45               0.95
hexapdf        3085.5       1.1         2102.4          2.7         2161.6           1.1         1.47              1.30               1.43               1.29
liquid-render  195.0        0.7         116.3           1.4         115.7            1.2         1.68              1.31               1.69               1.30
mail           174.8        0.2         151.7           0.2         152.0            0.1         1.15              0.81               1.15               0.97
optcarrot      6327.6       0.8         3454.5          0.7         3526.3           0.7         1.83              1.79               1.79               1.75
psych-load     2419.1       0.0         1899.3          0.0         1911.3           0.0         1.27              1.27               1.27               1.26
railsbench     2769.1       0.6         2075.8          1.5         2033.3           1.4         1.33              1.21               1.36               1.19
rubykon        12655.1      0.4         6887.4          1.0         6889.8           0.3         1.84              1.86               1.84               1.86
-------------  -----------  ----------  --------------  ----------  ---------------  ----------  ----------------  -----------------  -----------------  ------------------
```

There's more of a difference with call threshold 10 vs 500. 

The erubi_rails and mail benchmarks had the worst 1st iteration performance, and their 1st iteration performance is less bad (closer to 1.0, so closer to the interpreter) with call threshold 500.

After warmup, erubi_rails performs a little better, but railsbench performs a little bit worse 😢 

So, not sure what to conclude from this. I think we could bump the default threshold a bit with minimal repercussions, might make sense to make the default 20 or 50, but this also highlights that just having a call threshold is also a pretty limited way to decide which methods to JIT. Ideally, eventually we might want to more intelligently rank methods by how often they run.

**Comment by maximecb** [comment]
I had some additional thoughts on the issue: it seems like the call threshold heuristic is too basic to really tell us when a program is done booting, so to speak. Potentially, what we could use multiple simple heuristics to estimate when we are done booting. 

This is just an idea, but I'm assuming that during boot, a program will frequently start executing new functions and loading new modules. It might be possible to keep track of how recently the last function was first executed (or when the last module was loaded). Essentially, waiting until the rate at which new functions or modules are loaded starts to quiet down before activating YJIT.

**Comment by noahgibbs** [comment]
That sounds like a good idea. It would be a good way to check for the failure case -- common in Ruby -- where a program is doing something weird and we don't necessarily want to try to optimise it. For instance, a Rails server in development mode will just keep recompiling everything and YJIT won't help it.

Though it would also have false-positives like people using old-style structs that would constantly define new methods (or anything similar) where YJIT would help overall, but the program never really stops creating new code. I think that's reasonable, though -- those are already known performance sinkholes.

**Comment by k0kubun** [comment]
To me as a Ruby user, only two things matter in unit tests performance:

1. Time spent on running one or a few test cases
2. Total time spent on running all tests on CI

I'm not sure how bad we currently are at (1). I suspect majority of calls are not JITed anyway in this workload, but we should start measuring it somewhere to "make sure we don't make the performance of unit tests worse". Also, the bottleneck of (1) is typically the boot process, so once you cache the boot process with spring or bootsnap correctly, speeding up the rest might not be a high priority.

(2) is probably more important because it typically ends up taking a longer time than (1) in a whole development cycle even if you run (1) so many times locally. And fortunately, (2) might be a better workload for YJIT than (1) because some parts could be called so many times. Theoretically, if we compile only ISEQs whose speedup multiplied by the number of JITed calls is larger than its compilation time, we can not only avoid a slowdown but also make it faster than with the interpreter.

> Should we set up a unit test benchmark? Should we simply use first iteration time on the benchmarks we already have as a proxy?

Improving the first iteration time would likely improve unit tests workloads, so that sounds like a good proxy.

If we want to set up a unit test benchmark in yjit-bench, to confirm the actual numbers for (1), I suppose each iteration needs to run a separate child process and execute tests only once per test case. Most yjit-bench scripts currently run almost the same thing multiple times in each iteration, but a single test code should not be executed multiple times in the same process.

For (2), I don't think we need a proxy or a benchmark harness; we can directly look at the actual number, the total time spent on a test step on CI. For us, if we run Core or SFR with YJIT enabled on some CI, it'd be nice to monitor it. For the community, the CI of rails/rails (`rubylang/ruby:latest` vs `yjit:rubylang/ruby:latest`) might be a nice non-proprietary workload.

**Comment by maximecb** [comment]
We've bumped the call threshold to 30 by default, which should help. Going to close this for now as it's not a high priority and I think we're doing ok. Might still be nice to have a unit test benchmark in the future but we can revisit this later.
